### PR TITLE
Support multiple fonts adding an option to add extra fonts.  Add fontname support for clef and meter symbols. 

### PR DESCRIPTION
**Pull request by fernandoherreradelasheras**
See issue #3575

### CODE REVIEW

**Comment by lpugin** [comment]
Thanks. Could you make your font `Patriarca` available somewhere?

**Comment by lpugin** [comment]
You need to run clang-format. See https://book.verovio.org/contributing/guidelines.html#formatting

**Comment by fernandoherreradelasheras** [comment]
> Thanks. Could you make your font `Patriarca` available somewhere?

Sure!

Patriarca font is described in [^1]. It is based on Capitan font from another project (described in [^2]). The version I am using was included in the source code for the MuRET project [^3], however, the public repository with MuRET code [^4] was removed from github some months ago. There is still a public repository with one of the libraries developed for MuRET that has a copy of Patriarca font:

https://github.com/davidrizo/im3/tree/master/imcore/src/main/resources/fonts/patriarca

The font files don't have any copyright notice so I am not sure about the license for them (MuRET code itself was licensed using a mix of Apache and GPLv3).



[^1]: https://grfia.dlsi.ua.es/repositori/grfia/pubs/403/1953-8089-1-PB.pdf
[^2]: https://anuariomusical.revistas.csic.es/index.php/anuariomusical/article/view/221/221
[^3]: https://muret.dlsi.ua.es/muret/
[^4]: https://github.com/davidrizo/muret

**Comment by ahankinson** [comment]
Maybe @davidrizo can clarify the license question?

**Comment by fernandoherreradelasheras** [comment]
> You need to run clang-format. See https://book.verovio.org/contributing/guidelines.html#formatting

Done!

File: include/vrv/options.h [hunk 1]
```diff
@@ -686,6 +686,7 @@ class Options {
     OptionDbl m_extenderLineMinSpace;
     OptionDbl m_fingeringScale;
     OptionString m_font;
+    OptionArray m_addCustomFont;
```
Comment (lpugin) [inline include/vrv/options.h:689]: Rename to m_fontCustom

**Comment by fernandoherreradelasheras** [comment]
> Could you comment on what is hacky and patchy?

 When loading a font :
`LoadFont(const std::string &fontName, bool withFallback = true, bool buildNameTable = false);`
these semantics are confusing. We explicitly indicate in the call that we want to build the global name table from it and that we don't want to use a fallback font for any missing glyphs. However which font acts as fallback is implicitly indicated by the order of loading. Maybe we could add a separate call `SetFallbackFont` or add a `isFallbackFont` parameter to `LoadFont` .

About the withFallback parameter I am not sure, because it is not exposed via any option. Without being exposed it's just a policy . What should be this? verovio distributed fonts are expected to have all glyphs and they are safe to use without fallback? then any user added font (included if set as default) should be assumed that need fallback? Is there a use case when people might want to not render a all missing glyphs from they selected font?

Then the data structures I used for keeping several font maps are neither the most optimal or the most straightforward to follow: a map with file path as key and (glyph, id) struct as value, where id is calculated on insertion based on the number of glyphs loaded before for this same code + another map to keep that count. This was motivated because I thought maintaining svg output from different versions of verovio for the same input as similar as possible across different versions of verovio was a desirable goal (and this implementation was mapped around trying to keep the same format of the ids as SmuflCode-VariableSuffix for all the non-new cases). But then I though I was overthinking this and maybe that goal doesn't make that much sense in this case (with the variable suffix a diff between two runs of the same version is already not useful and only maybe some xslt filters may rely on this format).

Finally, code using different fonts does:
```
Resources &resources = m_doc->GetResourcesForModification();
resources.SetCurrentFont(customFont);
```
and maybe we should move the current font somewhere else because we are not really modifying the resources themselves just choosing one of the for the next operation.


Those were my main concerns about this implementation plus, of course, any other thing I might have missed.

**Comment by davidrizo** [comment]
> Maybe @davidrizo can clarify the license question?

Hi @fernandoherreradelasheras, @ahankinson 

Currently, Patriarca is being used inside MuRET to render the agnostic encoding. 

We have not yet finished the font to make it compatible with Verovio. There are always more urgents things to do... We have done some experiments with it, but elements such as the ligature are not working well yet, and some glyphs need to be corrected. Our plan is to have it ready for the summer, with the new version of MuRET.

It will be made publicly available, I don't know which license yet, it's possible that we'll use SIL Open Font License.

@fernandoherreradelasheras You can contact me by mail to know more details if you need.

**Pull request merged**